compass/reports.py

```diff
@@ -1,138 +1,124 @@
-    report.raise_for_status()
-
-    if "compass.scouts.org.uk%2fError.aspx|" in report.text:
-        raise CompassReportError("Compass Error!")
-
-    logger.info("Exporting report")
-    export_url_path, export_url_params = get_report_export_url(report_page.text)
-    csv_export = logon._get(f"{Settings.base_url}/{export_url_path}", params=export_url_params)
-
-    # TODO Debug check
-    logger.info("Saving report")
-    time_string = datetime.datetime.now().replace(microsecond=0).isoformat().replace(":", "-")  # colons are illegal on windows
-    filename = f"{time_string} - {logon.cn} ({logon.current_role}).csv"
-    Path(filename).write_bytes(csv_export.content)  # TODO Debug check
-
-    logger.debug(len(csv_export.content))
-    logger.info("Report Saved")
-
-    return csv_export.content
+TYPES_REPORTS = Literal[
+    "Region Member Directory",
+    "Region Appointments Report",
+    "Region Permit Report",
+    "Region Disclosure Report",
+    "Region Training Report",
+    "Region Disclosure Management Report",
+]
+
+
+class ReportTypes(enum.IntEnum):
+    region_member_directory = 37
+    region_appointments_report = 52
+    region_permit_report = 72
+    region_disclosure_report = 76
+    region_training_report = 84
+    region_disclosure_management_report = 100
+
+
+class Reports:
+    def __init__(self, session: Logon):
+        """Constructor for Reports."""
+        self._scraper = ReportsScraper(session.s)
+        self._scraper._get = session._get  # massively hacky but we need to send the jk_hash stuff through
+        self.session: Logon = session
+
+    def get_report(self, report_type: TYPES_REPORTS) -> bytes:
+        """Exports report as CSV from Compass.
+
+        Exporting a report is of course surprisingly complicated. The process
+        has four major steps, as follows:
+
+        1. Get a token for generating the report from the Compass backend. This
+            also validates that the report exists and that the user is
+            authenticated to access it.
+        2. If successful in obtaining a report token, get the initial report
+            page. The token is the (relative) URL here, and the report page
+            contains further needed information, such as the export URL and
+            location data for a full export.
+            (Compass does not include all organisational units in reports by
+            default, and to export all data for a given unit and downwards, we
+            need to add in these missing/unset levels manually).
+        3. We update report configuration data (sent as form data), and check
+            that we are not in an error state.
+        4. We extract the export URL, download the content and save to disk.
+
+        Pitfalls to be aware of in this process include that:
+        - Compass checks user-agent headers in some parts of the process
+            (TODO pinpoint which exactly)
```

Comment:
Interesting, is this simply checking for a valid browser UA or that the token request/download is the same UA? As you say, one to explore.

Reply:
It doesn't like requests' default user agent (python-requests/2.25.1, where 2.25.1 is the version) - the javascript evidences user-agent sniffing for front-end stuff so I wouldn't be suprised if it does it still backend - but just doing a stripped down version of the FF user-agent seemed to work. 

Need to track down exactly what breaks though for future ref & potential refactoring!